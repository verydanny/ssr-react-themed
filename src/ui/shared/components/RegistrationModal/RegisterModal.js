import React, { Component } from 'react'
import { themed } from 'react-themed'
import cn from 'classnames'
import autobind from 'autobind-decorator'
import { Modal } from '@rentpath/react-ui-core'
import { Icon } from 'ui/shared/components/Icon'
import close from 'ui/shared/components/Icon/svgs/global/close.svg'
import PropTypes from 'prop-types'
import Spinner from 'ui/shared/components/Spinner'
import UserModalForm from 'ui/shared/components/UserModal/components/UserModalForm'
import OauthButtons from 'ui/shared/components/UserModal/components/OauthButtons'
import { FIELD_MAP } from './const'

@themed('*')
export default class RegisterModal extends Component {

  static propTypes = {
    isOpen: PropTypes.bool,
    onClose: PropTypes.func,
    theme: PropTypes.object,
    isCountryUS: PropTypes.bool,
  }

  static defaultProps = {
    theme: {},
    isCountryUS: false,
  }

  constructor(props) {
    super(props)
    const { isCountryUS } = this.props
    this.state = {
      isFetching: false,
      fieldStatus: {}, // This will be used when validation is wired up
      rememberMe: true,
      optInNewsletter: isCountryUS,
    }
  }

  @autobind
  onSubmit() {
    /* TODO: should be wired up later and remove console  */
    console.log('This is the form data', this.state) // eslint-disable-line no-console
  }

  get header() {
    const { theme } = this.props
    return (
      <div>
        <div className={theme.RegisterModal_SignUpMessage}>
          Welcome, Sign Up For Free
        </div>
        <div>
          Creating an account allows you to acccess
          your saved and contacted properties on any device.
        </div>
      </div>
    )
  }

  /**
   * Finish this during the wire up story.
   * TODO
   * Error handling from API as validation is done via Zutron.
   * Wire up callback success and failure actions
   * Complete spinner
   */
  get submitContent() {
    const { theme } = this.props

    if (this.state.isFetching) {
      return <Spinner className={theme.Spinner} />
    }

    return 'Sign Up'
  }

  /**
   * TODO
   * Refactor first name and last name css to stack on smaller screens for mobile
   */
  get registrationFields() {
    const { theme, isCountryUS } = this.props
    const { fieldStatus, rememberMe, optInNewsletter } = this.state

    const allFields = [
      {
        name: 'firstName',
        type: 'text',
        modal: 'register',
        placeholder: 'First Name',
        'data-tag_item': 'firstName',
        autoComplete: 'given-name',
        onChange: this.handleInputChange,
        className: theme.RegisterModal_FirstName,
        ...fieldStatus.firstName,
      },
      {
        name: 'lastName',
        type: 'text',
        modal: 'register',
        placeholder: 'Last Name',
        'data-tag_item': 'lastName',
        autoComplete: 'family-name',
        onChange: this.handleInputChange,
        className: theme.RegisterModal_LastName,
        ...fieldStatus.lastName,
      },
      {
        name: 'email',
        placeholder: 'Email Address',
        modal: 'register',
        'data-tag_item': 'emailAddress',
        onChange: this.handleInputChange,
        autoComplete: 'email',
        className: theme.RegisterModal_Email,
        ...fieldStatus.email,
      },
      {
        name: 'password',
        type: 'password',
        modal: 'register',
        'data-tag_item': 'password',
        placeholder: 'Password',
        onChange: this.handleInputChange,
        autoComplete: 'off',
        ...fieldStatus.password,
      },
      {
        name: 'passwordConfirmation',
        type: 'password',
        modal: 'register',
        'data-tag_item': 'passwordConfirmation',
        placeholder: 'Confirm Password',
        onChange: this.handleInputChange,
        autoComplete: 'off',
      },
      {
        name: 'submit',
        modal: 'register',
        'data-tag_item': 'signup_button',
        children: this.submitContent,
        className: theme.ButtonRaised,
      },
      {
        name: 'rememberMe',
        type: 'checkbox',
        modal: 'register',
        'data-tag_item': 'checkbox',
        label: 'Keep me signed in',
        checked: rememberMe,
        onChange: this.handleInputChange,
        className: theme.RegisterModal_Checkbox,
      },
      {
        name: 'optInNewsletter',
        type: 'checkbox',
        modal: 'register',
        'data-tag_item': 'checkbox',
        label: 'Simplify my search with helpful tips and rental recommendations.',
        checked: optInNewsletter,
        onChange: this.handleInputChange,
        className: theme.RegisterModal_Checkbox,
      },
    ]

    return allFields.filter(field => !(field.name === 'optInNewsletter' && isCountryUS))
  }

  get footer() {
    const { theme } = this.props

    return (
      <div className={theme.RegisterModal_Footer}>
        <p>
          Already have an Acount?
        </p>
        <button
          className={cn(theme.RegisterModal_AccountLink, theme.RegisterModal_Button)}
        >
          Sign In
        </button>
      </div>
    )
  }

  @autobind
  handleInputChange(event) {
    const target = event.target
    const value = target.type === 'checkbox' ? target.checked : target.value

    this.setState({ [target.name]: value })
  }

  @autobind
  closeButton({ onClick }) {
    const { theme } = this.props

    return (
      <button
        tabIndex={0}
        className={cn(theme.RegisterModal_CloseButton, theme.RegisterModal_Button)}
        onClick={onClick}
        data-tag_item="x"
        data-tid="sign-in-close"
      >
        <Icon svgs={{ close }} />
      </button>
    )
  }

  render() {
    const {
      isOpen,
      onClose,
      theme,
    } = this.props

    return (
      <div
        className={cn(theme.LeadModal, theme.RegisterModal)}
        data-tag_section="register-modal"
      >
        <Modal
          CloseButton={this.closeButton}
          isOpen={isOpen}
          onClose={onClose}
          data-tid="register-modal"
        >
          <div>
            {this.header}
            <OauthButtons />
            <hr className={theme.RegisterModal_Orline} />
            <UserModalForm
              fields={this.registrationFields}
              onSubmit={this.onSubmit}
              fieldMap={FIELD_MAP}
              displayName="RegisterModalForm"
            />
            {this.footer}
          </div>
        </Modal>
      </div>
    )
  }
}
