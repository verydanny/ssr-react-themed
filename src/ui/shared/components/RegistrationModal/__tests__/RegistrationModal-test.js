import React from 'react'
import { render } from 'enzyme'
import RegisterModal from '../RegisterModal'

describe('RegisterModal', () => {
  describe('Modal is open', () => {
    const props = {
      onClose: () => true,
      isOpen: true,
      isCountryUS: false,
    }
    const wrapper = render(<RegisterModal {...props} />)

    describe('props.isCountryUS', () => {
      it('hides checkbox when true', () => {
        const component = render(<RegisterModal {...props} isCountryUS />)
        expect(
          component.find('[data-tid="user-modal-form-optInNewsletter"]').length
        ).toBe(0)
      })

      it('disabled checkbox when false', () => {
        const component = render(<RegisterModal {...props} />)
        expect(
          component.find('[data-tid="user-modal-form-optInNewsletter"]').prop('checked')
        ).toBeFalsy()
      })
    })

    describe('Modal tests', () => {
      it('matches snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
      })
    })
  })
})
