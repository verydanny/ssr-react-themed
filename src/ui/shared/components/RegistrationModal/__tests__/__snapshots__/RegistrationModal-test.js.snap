// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterModal Modal is open Modal tests matches snapshot 1`] = `

<div role="presentation"
     class
>
  <div class
       data-tid="register-modal"
  >
    <button tabindex="0"
            class
            data-tag_item="x"
            data-tid="sign-in-close"
    >
    </button>
    <div>
      <div>
        <div>
          <div>
            Welcome, Sign Up For Free
          </div>
          <div>
            Creating an account allows you to acccess your saved and contacted properties on any device.
          </div>
        </div>
        <div>
          <a class
             href="#"
          >
            <div style="height:50px"
                 class="lazyload-placeholder"
            >
            </div>
            <p>
              Sign in with Facebook
            </p>
          </a>
          <a class
             href="#"
          >
            <div style="height:50px"
                 class="lazyload-placeholder"
            >
            </div>
            <p>
              Sign in with Google
            </p>
          </a>
        </div>
        <hr>
        <form action="#"
              method="POST"
              novalidate
              class
        >
          <div class>
            <div class>
              <label data-tid="label"
                     for="RegisterModalForm-0-firstName"
                     class
              >
                First Name
              </label>
              <input type="text"
                     id="RegisterModalForm-0-firstName"
                     data-tid="user-register-form-firstName"
                     name="firstName"
                     modal="register"
                     data-tag_item="firstName"
                     autocomplete="given-name"
                     class
              >
            </div>
          </div>
          <div class>
            <div class>
              <label data-tid="label"
                     for="RegisterModalForm-1-lastName"
                     class
              >
                Last Name
              </label>
              <input type="text"
                     id="RegisterModalForm-1-lastName"
                     data-tid="user-register-form-lastName"
                     name="lastName"
                     modal="register"
                     data-tag_item="lastName"
                     autocomplete="family-name"
                     class
              >
            </div>
          </div>
          <div class>
            <div class>
              <label data-tid="label"
                     for="RegisterModalForm-2-email"
                     class
              >
                Email Address
              </label>
              <input type="text"
                     id="RegisterModalForm-2-email"
                     data-tid="user-register-form-email"
                     name="email"
                     modal="register"
                     data-tag_item="emailAddress"
                     autocomplete="email"
                     class
              >
            </div>
          </div>
          <div class>
            <div class>
              <label data-tid="label"
                     for="RegisterModalForm-3-password"
                     class
              >
                Password
              </label>
              <input type="password"
                     id="RegisterModalForm-3-password"
                     data-tid="user-register-form-password"
                     name="password"
                     modal="register"
                     data-tag_item="password"
                     autocomplete="off"
                     class
              >
            </div>
          </div>
          <div class>
            <div class>
              <label data-tid="label"
                     for="RegisterModalForm-4-passwordConfirmation"
                     class
              >
                Confirm Password
              </label>
              <input type="password"
                     id="RegisterModalForm-4-passwordConfirmation"
                     data-tid="user-register-form-passwordConfirmation"
                     name="passwordConfirmation"
                     modal="register"
                     data-tag_item="passwordConfirmation"
                     autocomplete="off"
                     class
              >
            </div>
          </div>
          <button data-tid="user-register-form-submit"
                  id="RegisterModalForm-5-submit"
                  name="submit"
                  modal="register"
                  data-tag_item="signup_button"
                  class
          >
            Sign Up
          </button>
          <div class>
            <label data-tid="label"
                   for="RegisterModalForm-6-rememberMe"
                   class
            >
              Keep me signed in
            </label>
            <input type="checkbox"
                   id="RegisterModalForm-6-rememberMe"
                   data-tid="user-register-form-rememberMe"
                   name="rememberMe"
                   modal="register"
                   data-tag_item="checkbox"
                   checked
                   class
            >
          </div>
          <div class>
            <label data-tid="label"
                   for="RegisterModalForm-7-optInNewsletter"
                   class
            >
              Simplify my search with helpful tips and rental recommendations.
            </label>
            <input type="checkbox"
                   id="RegisterModalForm-7-optInNewsletter"
                   data-tid="user-register-form-optInNewsletter"
                   name="optInNewsletter"
                   modal="register"
                   data-tag_item="checkbox"
                   class
            >
          </div>
        </form>
        <div>
          <p>
            Already have an Acount?
          </p>
          <button class>
            Sign In
          </button>
        </div>
      </div>
    </div>
  </div>
</div>

`;
